# Boston Housing Dataset Example

Most of this code is included in Chapter 8 of [Data Science for Mathematicians](https://ds4m.github.io/).  We convert it to notebook form here so that you can see the output and explore it interactively online yourself.  To run the notebook in a new Google Colab project, click here:

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ds4m/ds4m.github.io/blob/master/chapter-8-resources/boston-dataset-example.ipynb)

## Step 1: Obtain data

The Boston housing dataset is built into scikit-learn, so we can import it easily, as follows.

In [1]:
from sklearn.datasets import load_boston
boston = load_boston()

But the `boston` object created this way is a conglomeration of several sub-objects and not ready to be printed in a human-readable way, so we organize it as follows.

## Step 2: Create a feature-target dataset

We extract the features and target into separate variables and inspect their contents.

In [2]:
import pandas as pd
features = pd.DataFrame(
    data=boston.data,
    columns=boston.feature_names)
target = boston.target

The features are a pandas DataFrame, the first few rows of which are shown here.

In [3]:
features.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

How many rows are there, actually?

In [4]:
len( features )

506

The target is a NumPy array that can be viewed as another column in the same dataset, as shown here.

In [5]:
target

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

## Step 3: Split into training and test datasets

We will use 80% of the data for training and then test our model on the 20% held out for that purpose.  Scikit-learn contains a function that will randomly split the dataset for us into training and test sets.  We add the `random_state` parameter to specify a random number seed, thus guaranteeing reproducibility of the same results if you re-run this notebook later.

In [6]:
from sklearn.model_selection import train_test_split
(X_training, X_test, y_training, y_test) = \
    train_test_split(features,target,train_size=0.8,random_state=1)

Let's verify that the split produced objects of the appropriate sizes.

In [7]:
len( X_training ), len( X_test ), len( y_training ), len( y_test )

(404, 102, 404, 102)

Since 404 is almost exactly 80% of the original 506, it looks like this has worked correctly.

## Step 4: Create a model from the training dataset

We use scikit-learn's Pipeline object to compose two steps in sequence:  First, select the five best features to use for prediction, and second, use those five features to fit a linear model to the training data.

In [8]:
### Step 4: Create a model from the training dataset
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SelectKBest
from sklearn.linear_model import LinearRegression
estimator = Pipeline([
    ('select', SelectKBest(k=5)),
    ('model',  LinearRegression())
])
fit_model = estimator.fit(X=X_training , y=y_training)

Let's say we'd like to see which features were selected.  We can ask the first step in the pipeline to show us its results.

In [9]:
features.columns[ fit_model[0].get_support() ]

Index(['CRIM', 'NOX', 'RM', 'AGE', 'LSTAT'], dtype='object')

And if we want to see the coefficients the model assigned to each of those variables, we can ask teh second step in the pipeline for its results.

In [10]:
fit_model[1].intercept_, fit_model[1].coef_

(2.9524367266279405,
 array([-0.09549911, -4.08891308,  4.56355544,  0.02161194, -0.61759647]))

The resulting model is therefore approximately the following.
$$ 2.952 - 0.095\text{CRIM} - 4.089\text{NOX} + 4.564\text{RM} + 0.022\text{AGE} - 0.618\text{LSTAT} $$

## Step 5: Score the model using the test set

We compute the root mean squared error of the model on the test set.

In [11]:
predictions = fit_model.predict(X=X_test)
from sklearn.metrics import mean_squared_error
mean_squared_error(y_test, predictions)**0.5

5.630885425217404